encodings/dict/src/builders/primitive.rs

```diff
@@ -10,69 +10,115 @@ use vortex_array::validity::Validity;
 use vortex_array::{Array, ArrayRef, ToCanonical};
 use vortex_buffer::BufferMut;
 use vortex_dtype::{NativePType, Nullability, PType};
-use vortex_error::{VortexExpect, VortexResult, vortex_bail};
+use vortex_error::{VortexExpect, VortexResult, vortex_bail, vortex_panic};
 
+use super::DictConstraints;
 use crate::builders::DictEncoder;
 
-impl<T: NativePType> PrimitiveDictBuilder<T>
+pub fn primitive_dict_builder<T: NativePType>(
+    nullability: Nullability,
+    constraints: &DictConstraints,
+) -> Box<dyn DictEncoder>
 where
     NativeValue<T>: Hash + Eq,
 {
-    pub fn new(nullability: Nullability, max_dict_bytes: usize) -> Self {
+    // bound constraints with the cardinality of the primitive type
+    let max_possible_len = (constraints.max_len as u64).min(match T::PTYPE.bit_width() {
+        8 => u8::MAX as u64,
+        16 => u16::MAX as u64,
+        32 => u32::MAX as u64,
+        64 => u64::MAX as u64,
+        width => vortex_panic!("invalid bit_width: {width}"),
+    });
+    match max_possible_len {
+        max if max <= u8::MAX as u64 => {
+            Box::new(PrimitiveDictBuilder::<T, u8>::new(nullability, constraints))
+        }
+        max if max <= u16::MAX as u64 => Box::new(PrimitiveDictBuilder::<T, u16>::new(
+            nullability,
+            constraints,
+        )),
+        max if max <= u32::MAX as u64 => Box::new(PrimitiveDictBuilder::<T, u32>::new(
+            nullability,
+            constraints,
+        )),
+        _ => Box::new(PrimitiveDictBuilder::<T, u64>::new(
+            nullability,
+            constraints,
+        )),
+    }
+}
+
+impl<T: NativePType, Code> PrimitiveDictBuilder<T, Code>
+where
+    NativeValue<T>: Hash + Eq,
+    Code: Unsigned + AsPrimitive<usize> + NativePType,
+{
+    pub fn new(nullability: Nullability, constraints: &DictConstraints) -> Self {
+        let max_dict_len = constraints
+            .max_len
+            .min(constraints.max_bytes / T::PTYPE.byte_width());
         Self {
             lookup: HashMap::with_hasher(FxBuildHasher),
             values: BufferMut::<T>::empty(),
             nullability,
-            max_dict_len: max_dict_bytes / T::PTYPE.byte_width(),
+            max_dict_len,
         }
     }
 
     #[inline]
-    fn encode_value(&mut self, v: T) -> u64 {
+    fn encode_value(&mut self, v: T) -> Option<Code> {
```

Comment:
previously, when we were only checking max size, we were doing the check after inserting the encoded value, because we didn't needed to be strict. Now with max len I want to strictly enforce the constraints so now we do check before encoding the value in both builders